app/models/article.rb

```diff
@@ -27,20 +27,48 @@ class Article < ApplicationRecord
 
   scope :global_newsfeed, -> { published.order(published_at: :desc) }
 
-  scope :with_tag, ->(tag_id) { joins(:article_tags).where(article_tags: { tag: tag_id }) }
-
   scope :with_partner_tag, lambda { |tag_id|
     joins('left outer join article_partners on articles.id=article_partners.article_id')
     .joins('left outer join partner_tags on article_partners.partner_id = partner_tags.partner_id')
     .where('partner_tags.tag_id = ?', tag_id)
   }
 
+  scope :with_tags, lambda { |tag_ids|
+    joins(:article_tags).where(article_tags: { tag: tag_ids })
+  }
+
   scope :for_site, lambda { |site|
-    site_neighbourhood_ids = site.owned_neighbourhoods.pluck(:id)
+    scope = all
 
-    joins(partners: [:address])
-      .where(address: { neighbourhood_id: site_neighbourhood_ids } )
-      .distinct
+    site_neighbourhood_ids = site.owned_neighbourhoods.pluck(:id)
+    site_tag_ids = site.tags.pluck(:id)
+    return scope if site_neighbourhood_ids.empty? && site_tag_ids.empty?
+
+    where_fragments = []
+    where_params = []
+
+    # articles by neighbourhood
+    if site_neighbourhood_ids.any?
```

Comment:
im wondering if its possible at all to wrap these? only because atm the cognitive load of reading this block with the long sql statements is very high